### PR TITLE
feat: additional kl metrics

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
# What does this PR do ?

Add metrics:
image
image

* https://wandb.ai/nvidia/grpo-dev-zhiyul/runs/kgjjunme?nw=poxtgkh2ggc
* https://wandb.ai/nvidia/zhiyul-nano-v3/runs/yeulq88w?nw=pty704q4dt

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Added three divergence metrics (gen_kl, policy_kl, and js_divergence) to track differences between model generation and training distributions during reinforcement learning training.

* **Documentation**
  * Added comprehensive guide explaining KL Divergence metrics, including definitions, interpretation guidance, and practical examples for monitoring model behavior.

### CODE REVIEW

File: docs/guides/grpo.md [hunk 1]
```diff
@@ -258,6 +258,35 @@ $$
...
+   logp_policy (from Mcore):      -15.xxx
+```
+Assuming other tokens have near-zero divergence, this single token's metrics are:
+
+* `gen_kl`: exp(-15 + 5) - (-15 + 5) - 1 = 9, or moderate mismatch
+* `policy_kl`: exp(-5 + 15) - (-5 + 15) - 1 = 22015, servere mismatch dominiting the metrics.
+* `js_divergence`: close to `gen_kl` since the mixture distribution (-5.69) heavily weights higher probability, i.e. logp_gen in this example
+
+Ideally, all KL Divergence should be close to 0 with a tolerance threshold of around 1e-3, while we need to investigate case by case if there's spikes in some metrics.
+
```
[the remark this answers is not in the document]
Comment (ZhiyuLi-Nvidia) [reply docs/guides/grpo.md]: Done.

File: nemo_rl/algorithms/loss_functions.py [hunk 2]
```diff
@@ -168,6 +168,50 @@ def __call__(
...
+            torch.exp(generation_logprobs - log_mixture)
+            - (generation_logprobs - log_mixture)
+            - 1
+        )
+
+        js_divergence = masked_mean(
+            0.5 * kl_prev_to_mixture + 0.5 * kl_gen_to_mixture,
+            mask,
+            global_normalization_factor=global_valid_toks,
+        ).item()
```
[the remark this answers is not in the document]
Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py]: Done.

File: nemo_rl/algorithms/loss_functions.py [hunk 3]
```diff
@@ -369,6 +413,9 @@ def __call__(
                 "probs_ratio_clamped": probs_ratio_clamped,
                 "kl_penalty": kl.item() / self.reference_policy_kl_penalty if kl else 0,
                 "token_mult_prob_error": mult_prob_error,
+                "gen_kl": gen_kl,
+                "policy_kl": policy_kl,
+                "js_divergence": js_divergence,
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
                "gen_kl_error": gen_kl,
                "policy_kl_error": policy_kl,
                "js_divergence_error": js_divergence,
```
maybe these should have this suffix to not get confused with the KL that's part of the objective

Comment (terrykong) [reply nemo_rl/algorithms/loss_functions.py]: and i guess the doc should be changed to reflect the new names

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py]: Thank you for comments. Done both in implementation and doc.

**Review by terrykong — APPROVED** [review]

File: nemo_rl/algorithms/loss_functions.py [hunk 4]
```diff
@@ -168,6 +168,50 @@ def __call__(
             global_normalization_factor=global_valid_toks,
         ).item()
 
+        # gen-kl(kl(P_gen || P_train)) = torch.exp(log_ratio) - log_ratio - 1
+        # where log_ratio = prev_logprobs - generation_logprobs
+        gen_kl_error = masked_mean(
```
Comment (yuki-97) [inline nemo_rl/algorithms/loss_functions.py:173]: just curious, why we not just use `calculate_kl_penalty_joschu2020`? seems they should be the same.

FYI I'm adding other kl type in #1349, so it'll be easy to use the same kl type for the metrics if we could also use that function.

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py:173]: Thank you @yuki-97. I think the goal might be slightly different.

`calculate_kl_penalty_joschu2020` seems to measure `KL(logprobs_policy, logprobs_reference)` where **`logprobs_reference`** is from reference model. The penalty will ensure the training policy stay not too far away from the reference model to avoid catastrophic forgetting.  
This metrics is simply an error to measure off-policy gaps between logprobs_policy and **generation_policy** which should be the same ideally.

Comment (yuki-97) [reply nemo_rl/algorithms/loss_functions.py:173]: Got it, thanks for explaining, so my understanding is that the calculation is functionally the same but the goal is different.

I'm thinking to change `calculate_kl_penalty_joschu2020(logprobs_policy, logprobs_reference)` to `calculate_kl(logprobs, logprobs_reference)` to make it more clear in my PR later and will need your review that time.
After that I think it's just a util function to calculate the kl between two logprobs instead of specifying they are the logprobs from training model and reference model, so that it's reasonable for these metrics stuffs to also use the function, is it ok for you?

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/loss_functions.py:173]: Thank you @yuki-97. Sounds a good idea! Feel free to make it unit function to avoid repeating ourselves!

**Review by terrykong — APPROVED** [review]
